File: app/lib/models/processing_pipeline.dart
import 'package:json_annotation/json_annotation.dart';

import 'anti_alias_parameters.dart';
import 'chroma_fix_parameters.dart';
import 'color_correction_parameters.dart';
import 'crop_resize_parameters.dart';
import 'deband_parameters.dart';
import 'deblock_parameters.dart';
import 'descratch_parameters.dart';
import 'spotless_parameters.dart';
import 'stabilize_parameters.dart';
import 'dehalo_parameters.dart';
import 'dynamic_parameters.dart';
import 'chroma_denoise_parameters.dart';
import 'noise_reduction_parameters.dart';
import 'parameter_converter.dart';
import 'qtgmc_parameters.dart';
import 'sharpen_parameters.dart';
import 'subtitle_parameters.dart';

part 'processing_pipeline.g.dart';

/// Defines the type of each processing pass.
enum PassType {
  deinterlace,
  descratch,
  spotless,
  noiseReduction,
  chromaDenoise,
  dehalo,
  deblock,
  deband,
  sharpen,
  antiAlias,
  stabilize,
  colorCorrection,
  chromaFixes,
  cropResize,
  subtitles,
}

/// Extension to provide display names for pass types.
extension PassTypeExtension on PassType {
  String get displayName {
    switch (this) {
      case PassType.deinterlace:
        return 'Deinterlace';
      case PassType.descratch:
        return 'DeScratch';
      case PassType.spotless:
        return 'SpotLess';
      case PassType.noiseReduction:
        return 'Noise Reduction';
      case PassType.chromaDenoise:
        return 'Chroma Denoise';
      case PassType.dehalo:
        return 'Dehalo';
      case PassType.deblock:
        return 'Deblock';
      case PassType.deband:
        return 'Deband';
      case PassType.sharpen:
        return 'Sharpen';
      case PassType.antiAlias:
        return 'Anti-Aliasing';
      case PassType.stabilize:
        return 'Stabilize';
      case PassType.colorCorrection:
        return 'Color Correction';
      case PassType.chromaFixes:
        return 'Chroma Fixes';
      case PassType.cropResize:
        return 'Crop / Resize';
      case PassType.subtitles:
        return 'Subtitles';
    }
  }

  String get description {
    switch (this) {
      case PassType.deinterlace:
        return 'Deinterlace (QTGMC) or inverse telecine (IVTC)';
      case PassType.descratch:
        return 'Remove vertical scratches from scanned film';
      case PassType.spotless:
        return 'Remove dust, dirt, and temporal spots from film';
      case PassType.noiseReduction:
        return 'Reduce video noise and grain';
      case PassType.chromaDenoise:
        return 'Remove blotchy colour noise (CCD)';
      case PassType.dehalo:
        return 'Remove halo artifacts around edges';
      case PassType.deblock:
        return 'Remove compression block artifacts';
      case PassType.deband:
        return 'Remove color banding from gradients';
      case PassType.sharpen:
        return 'Sharpen edges and enhance detail';
      case PassType.antiAlias:
        return 'Smooth stair-stepping on diagonal edges';
      case PassType.stabilize:
        return 'Remove shake and weave from the picture';
      case PassType.colorCorrection:
        return 'Adjust brightness, contrast, and colors';
      case PassType.chromaFixes:
        return 'Fix chroma bleeding and crawl artifacts';
      case PassType.cropResize:
        return 'Crop borders and resize output';
      case PassType.subtitles:
        return 'Generate subtitles using Whisper AI';
    }
  }
}

/// Container for all processing pass parameters.
/// Defines the complete video processing pipeline.
@JsonSerializable(explicitToJson: true)
class ProcessingPipeline {
  /// Deinterlacing pass parameters (QTGMC).
  final QTGMCParameters deinterlace;

  /// DeScratch pass parameters (scratch removal).
  final DeScratchParameters descratch;

  /// SpotLess pass parameters (spot/dirt removal).
  final SpotLessParameters spotless;

  /// Noise reduction pass parameters.
  final NoiseReductionParameters noiseReduction;

  /// Chroma denoise pass parameters (CCD).
  final ChromaDenoiseParameters chromaDenoise;

  /// Dehalo pass parameters.
  final DehaloParameters dehalo;

  /// Deblock pass parameters.
  final DeblockParameters deblock;

  /// Deband pass parameters (f3kdb).
  final DebandParameters deband;

  /// Sharpening pass parameters.
  final SharpenParameters sharpen;
  final AntiAliasParameters antiAlias;
  final StabilizeParameters stabilize;

  /// Color correction pass parameters.
  final ColorCorrectionParameters colorCorrection;

  /// Chroma fix pass parameters.
  final ChromaFixParameters chromaFixes;

  /// Crop and resize pass parameters.
  final CropResizeParameters cropResize;

  /// Subtitle generation pass parameters (Whisper AI).
  final SubtitleParameters subtitles;

  const ProcessingPipeline({
    this.deinterlace = const QTGMCParameters(),
    this.descratch = const DeScratchParameters(),
    this.spotless = const SpotLessParameters(),
    this.noiseReduction = const NoiseReductionParameters(),
    this.chromaDenoise = const ChromaDenoiseParameters(),
    this.dehalo = const DehaloParameters(),
    this.deblock = const DeblockParameters(),
    this.deband = const DebandParameters(),
    this.sharpen = const SharpenParameters(),
    this.antiAlias = const AntiAliasParameters(),
    this.stabilize = const StabilizeParameters(),
    this.colorCorrection = const ColorCorrectionParameters(),
    this.chromaFixes = const ChromaFixParameters(),
    this.cropResize = const CropResizeParameters(),
    this.subtitles = const SubtitleParameters(),
  });

  /// Create a pipeline from legacy QTGMC-only parameters.
  factory ProcessingPipeline.fromLegacy(QTGMCParameters qtgmcParams) {
    return ProcessingPipeline(
      deinterlace: qtgmcParams,
      // Other passes disabled by default when migrating from legacy
      descratch: const DeScratchParameters(enabled: false),
      spotless: const SpotLessParameters(enabled: false),
      noiseReduction: const NoiseReductionParameters(enabled: false),
      chromaDenoise: const ChromaDenoiseParameters(enabled: false),
      dehalo: const DehaloParameters(enabled: false),
      deblock: const DeblockParameters(enabled: false),
      deband: const DebandParameters(enabled: false),
      sharpen: const SharpenParameters(enabled: false),
      antiAlias: const AntiAliasParameters(enabled: false),
      stabilize: const StabilizeParameters(enabled: false),
      colorCorrection: const ColorCorrectionParameters(enabled: false),
      chromaFixes: const ChromaFixParameters(enabled: false),
      cropResize: const CropResizeParameters(enabled: false),
      subtitles: const SubtitleParameters(enabled: false),
    );
  }

  /// Get the ordered list of enabled passes.
  List<PassType> get enabledPasses {
    final passes = <PassType>[];
    // Order: Crop first (pre-processing), then deinterlace, noise, dehalo, deblock, deband, sharpen, chroma, color, resize last
    if (cropResize.enabled && cropResize.cropEnabled) {
      passes.add(PassType.cropResize); // Pre-crop
    }
    if (deinterlace.enabled) {
      passes.add(PassType.deinterlace);
    }
    if (descratch.enabled) {
      passes.add(PassType.descratch);
    }
    if (spotless.enabled) {
      passes.add(PassType.spotless);
    }
    if (noiseReduction.enabled) {
      passes.add(PassType.noiseReduction);
    }
    // Chroma noise is removed with the rest of the denoising, before the
    // deband/sharpen steps that would otherwise amplify it.
    if (chromaDenoise.enabled) {
      passes.add(PassType.chromaDenoise);
    }
    if (dehalo.enabled) {
      passes.add(PassType.dehalo);
    }
    if (deblock.enabled) {
      passes.add(PassType.deblock);
    }
    if (deband.enabled) {
      passes.add(PassType.deband);
    }
    // Anti-aliasing before sharpening: sharpening stair-stepped edges makes
    // the stepping more visible, not less.
    if (antiAlias.enabled) {
      passes.add(PassType.antiAlias);
    }
    if (sharpen.enabled) {
      passes.add(PassType.sharpen);
    }
    if (chromaFixes.enabled) {
      passes.add(PassType.chromaFixes);
    }
    if (colorCorrection.enabled) {
      passes.add(PassType.colorCorrection);
    }
    // Stabilisation shifts the picture within the frame, so it runs last
    // before framing — a crop can then remove the edges it exposes.
    if (stabilize.enabled) {
      passes.add(PassType.stabilize);
    }
    if (cropResize.enabled && cropResize.resizeEnabled) {
      // Resize (post-processing) - if not already added for crop
      if (!passes.contains(PassType.cropResize)) {
        passes.add(PassType.cropResize);
      }
    }
    return passes;
  }

  /// Get count of enabled passes (includes subtitles).
  int get enabledPassCount {
    var count = 0;
    if (deinterlace.enabled) count++;
    if (descratch.enabled) count++;
    if (spotless.enabled) count++;
    if (noiseReduction.enabled) count++;
    if (chromaDenoise.enabled) count++;
    if (dehalo.enabled) count++;
    if (deblock.enabled) count++;
    if (deband.enabled) count++;
    if (sharpen.enabled) count++;
    if (antiAlias.enabled) count++;
    if (stabilize.enabled) count++;
    if (colorCorrection.enabled) count++;
    if (chromaFixes.enabled) count++;
    if (cropResize.enabled) count++;
    if (subtitles.enabled) count++;
    return count;
  }

  /// Get count of enabled video processing passes (excludes subtitles).
  int get videoPassCount {
    var count = 0;
    if (deinterlace.enabled) count++;
    if (descratch.enabled) count++;
    if (spotless.enabled) count++;
    if (noiseReduction.enabled) count++;
    if (chromaDenoise.enabled) count++;
    if (dehalo.enabled) count++;
    if (deblock.enabled) count++;
    if (deband.enabled) count++;
    if (sharpen.enabled) count++;
    if (antiAlias.enabled) count++;
    if (stabilize.enabled) count++;
    if (colorCorrection.enabled) count++;
    if (chromaFixes.enabled) count++;
    if (cropResize.enabled) count++;
    return count;
  }

  /// Check if a specific pass is enabled.
  bool isPassEnabled(PassType pass) {
    switch (pass) {
      case PassType.deinterlace:
        return deinterlace.enabled;
      case PassType.descratch:
        return descratch.enabled;
      case PassType.spotless:
        return spotless.enabled;
      case PassType.noiseReduction:
        return noiseReduction.enabled;
      case PassType.chromaDenoise:
        return chromaDenoise.enabled;
      case PassType.dehalo:
        return dehalo.enabled;
      case PassType.deblock:
        return deblock.enabled;
      case PassType.deband:
        return deband.enabled;
      case PassType.sharpen:
        return sharpen.enabled;
      case PassType.antiAlias:
        return antiAlias.enabled;
      case PassType.stabilize:
        return stabilize.enabled;
      case PassType.colorCorrection:
        return colorCorrection.enabled;
      case PassType.chromaFixes:
        return chromaFixes.enabled;
      case PassType.cropResize:
        return cropResize.enabled;
      case PassType.subtitles:
        return subtitles.enabled;
    }
  }

  /// Get summary string for a specific pass.
  String getPassSummary(PassType pass) {
    switch (pass) {
      case PassType.deinterlace:
        if (!deinterlace.enabled) return 'Off';
        if (deinterlace.method == DeinterlaceMethod.ivtc) return 'IVTC';
        return deinterlace.preset.displayName;
      case PassType.descratch:
        return descratch.summary;
      case PassType.spotless:
        return spotless.summary;
      case PassType.noiseReduction:
        return noiseReduction.summary;
      case PassType.chromaDenoise:
        return chromaDenoise.summary;
      case PassType.dehalo:
        return dehalo.summary;
      case PassType.deblock:
        return deblock.summary;
      case PassType.deband:
        return deband.summary;
      case PassType.sharpen:
        return sharpen.summary;
      case PassType.antiAlias:
        return antiAlias.summary;
      case PassType.stabilize:
        return stabilize.summary;
      case PassType.colorCorrection:
        return colorCorrection.summary;
      case PassType.chromaFixes:
        return chromaFixes.summary;
      case PassType.cropResize:
        return cropResize.summary;
      case PassType.subtitles:
        return subtitles.summary;
    }
  }

  ProcessingPipeline copyWith({
    QTGMCParameters? deinterlace,
    DeScratchParameters? descratch,
    SpotLessParameters? spotless,
    NoiseReductionParameters? noiseReduction,
    ChromaDenoiseParameters? chromaDenoise,
    DehaloParameters? dehalo,
    DeblockParameters? deblock,
    DebandParameters? deband,
    SharpenParameters? sharpen,
    AntiAliasParameters? antiAlias,
    StabilizeParameters? stabilize,
    ColorCorrectionParameters? colorCorrection,
    ChromaFixParameters? chromaFixes,
    CropResizeParameters? cropResize,
    SubtitleParameters? subtitles,
  }) {
    return ProcessingPipeline(
      deinterlace: deinterlace ?? this.deinterlace,
      descratch: descratch ?? this.descratch,
      spotless: spotless ?? this.spotless,
      noiseReduction: noiseReduction ?? this.noiseReduction,
      chromaDenoise: chromaDenoise ?? this.chromaDenoise,
      dehalo: dehalo ?? this.dehalo,
      deblock: deblock ?? this.deblock,
      deband: deband ?? this.deband,
      sharpen: sharpen ?? this.sharpen,
      antiAlias: antiAlias ?? this.antiAlias,
      stabilize: stabilize ?? this.stabilize,
      colorCorrection: colorCorrection ?? this.colorCorrection,
      chromaFixes: chromaFixes ?? this.chromaFixes,
      cropResize: cropResize ?? this.cropResize,
      subtitles: subtitles ?? this.subtitles,
    );
  }

  /// Toggle a pass on or off.
  ProcessingPipeline togglePass(PassType pass, bool enabled) {
    switch (pass) {
      case PassType.deinterlace:
        return copyWith(
          deinterlace: deinterlace.copyWith(enabled: enabled),
        );
      case PassType.descratch:
        return copyWith(
          descratch: descratch.copyWith(enabled: enabled),
        );
      case PassType.spotless:
        return copyWith(
          spotless: spotless.copyWith(enabled: enabled),
        );
      case PassType.chromaDenoise:
        return copyWith(
          chromaDenoise: chromaDenoise.copyWith(enabled: enabled),
        );
      case PassType.noiseReduction:
        return copyWith(
          noiseReduction: noiseReduction.copyWith(enabled: enabled),
        );
      case PassType.dehalo:
        return copyWith(
          dehalo: dehalo.copyWith(enabled: enabled),
        );
      case PassType.deblock:
        return copyWith(
          deblock: deblock.copyWith(enabled: enabled),
        );
      case PassType.deband:
        return copyWith(
          deband: deband.copyWith(enabled: enabled),
        );
      case PassType.sharpen:
        return copyWith(
          sharpen: sharpen.copyWith(enabled: enabled),
        );
      case PassType.antiAlias:
        return copyWith(
          antiAlias: antiAlias.copyWith(enabled: enabled),
        );
      case PassType.stabilize:
        return copyWith(
          stabilize: stabilize.copyWith(enabled: enabled),
        );
      case PassType.colorCorrection:
        return copyWith(
          colorCorrection: colorCorrection.copyWith(enabled: enabled),
        );
      case PassType.chromaFixes:
        return copyWith(
          chromaFixes: chromaFixes.copyWith(enabled: enabled),
        );
      case PassType.cropResize:
        return copyWith(
          cropResize: cropResize.copyWith(enabled: enabled),
        );
      case PassType.subtitles:
        return copyWith(
          subtitles: subtitles.copyWith(enabled: enabled),
        );
    }
  }

  /// Convert to a dynamic pipeline for schema-based processing.
  DynamicPipeline toDynamicPipeline() {
    return ParameterConverter.fromPipeline(this);
  }

  factory ProcessingPipeline.fromJson(Map<String, dynamic> json) =>
      _$ProcessingPipelineFromJson(json);

  Map<String, dynamic> toJson() => _$ProcessingPipelineToJson(this);
}

File: worker/src/models/processing_pipeline.rs
//! Processing pipeline containing all video processing passes.

use serde::{Deserialize, Serialize};

use super::{
    ChromaFixParameters, ColorCorrectionParameters, CropResizeParameters,
    DebandParameters, DeblockParameters, DehaloParameters, DeScratchParameters,
    SharpenParameters, NoiseReductionParameters, QTGMCParameters,
};

/// Defines the type of each processing pass.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
#[allow(dead_code)]
pub enum PassType {
    Deinterlace,
    DeScratch,
    NoiseReduction,
    Dehalo,
    Deblock,
    Deband,
    Sharpen,
    ColorCorrection,
    ChromaFixes,
    CropResize,
}

#[allow(dead_code)]
impl PassType {
    /// Get display name for the pass.
    pub fn display_name(&self) -> &'static str {
        match self {
            PassType::Deinterlace => "Deinterlace",
            PassType::DeScratch => "DeScratch",
            PassType::NoiseReduction => "Noise Reduction",
            PassType::Dehalo => "Dehalo",
            PassType::Deblock => "Deblock",
            PassType::Deband => "Deband",
            PassType::Sharpen => "Sharpen",
            PassType::ColorCorrection => "Color Correction",
            PassType::ChromaFixes => "Chroma Fixes",
            PassType::CropResize => "Crop / Resize",
        }
    }

    /// Get description for the pass.
    pub fn description(&self) -> &'static str {
        match self {
            PassType::Deinterlace => "Deinterlace (QTGMC) or inverse telecine (IVTC)",
            PassType::DeScratch => "Remove vertical scratches from scanned film",
            PassType::NoiseReduction => "Reduce video noise and grain",
            PassType::Dehalo => "Remove halo artifacts around edges",
            PassType::Deblock => "Remove compression block artifacts",
            PassType::Deband => "Remove color banding from gradients",
            PassType::Sharpen => "Sharpen edges and enhance detail",
            PassType::ColorCorrection => "Adjust brightness, contrast, and colors",
            PassType::ChromaFixes => "Fix chroma bleeding and crawl artifacts",
            PassType::CropResize => "Crop borders and resize output",
        }
    }
}

/// Container for all processing pass parameters.
/// Defines the complete video processing pipeline.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProcessingPipeline {
    /// Deinterlacing pass parameters (QTGMC).
    #[serde(default)]
    pub deinterlace: QTGMCParameters,

    /// DeScratch pass parameters (scratch removal).
    #[serde(default)]
    pub descratch: DeScratchParameters,

    /// Noise reduction pass parameters.
    #[serde(default)]
    pub noise_reduction: NoiseReductionParameters,

    /// Dehalo pass parameters.
    #[serde(default)]
    pub dehalo: DehaloParameters,

    /// Deblock pass parameters.
    #[serde(default)]
    pub deblock: DeblockParameters,

    /// Deband pass parameters (f3kdb).
    #[serde(default)]
    pub deband: DebandParameters,

    /// Sharpening pass parameters.
    #[serde(default)]
    pub sharpen: SharpenParameters,

    /// Color correction pass parameters.
    #[serde(default)]
    pub color_correction: ColorCorrectionParameters,

    /// Chroma fix pass parameters.
    #[serde(default)]
    pub chroma_fixes: ChromaFixParameters,

    /// Crop and resize pass parameters.
    #[serde(default)]
    pub crop_resize: CropResizeParameters,
}

impl Default for ProcessingPipeline {
    fn default() -> Self {
        Self {
            deinterlace: QTGMCParameters::default(),
            descratch: DeScratchParameters::default(),
            noise_reduction: NoiseReductionParameters::default(),
            dehalo: DehaloParameters::default(),
            deblock: DeblockParameters::default(),
            deband: DebandParameters::default(),
            sharpen: SharpenParameters::default(),
            color_correction: ColorCorrectionParameters::default(),
            chroma_fixes: ChromaFixParameters::default(),
            crop_resize: CropResizeParameters::default(),
        }
    }
}

#[allow(dead_code)]
impl ProcessingPipeline {
    /// Create a pipeline from legacy QTGMC-only parameters.
    pub fn from_legacy(qtgmc_params: &QTGMCParameters) -> Self {
        Self {
            deinterlace: qtgmc_params.clone(),
            descratch: DeScratchParameters { enabled: false, ..Default::default() },
            noise_reduction: NoiseReductionParameters { enabled: false, ..Default::default() },
            dehalo: DehaloParameters { enabled: false, ..Default::default() },
            deblock: DeblockParameters { enabled: false, ..Default::default() },
            deband: DebandParameters { enabled: false, ..Default::default() },
            sharpen: SharpenParameters { enabled: false, ..Default::default() },
            color_correction: ColorCorrectionParameters { enabled: false, ..Default::default() },
            chroma_fixes: ChromaFixParameters { enabled: false, ..Default::default() },
            crop_resize: CropResizeParameters { enabled: false, ..Default::default() },
        }
    }

    /// Get the ordered list of enabled passes.
    pub fn enabled_passes(&self) -> Vec<PassType> {
        let mut passes = Vec::new();

        // Order: Crop first (pre-processing), then deinterlace, noise, dehalo, deblock, deband, sharpen, chroma, color, resize last
        if self.crop_resize.enabled && self.crop_resize.crop_enabled {
            passes.push(PassType::CropResize); // Pre-crop
        }
        if self.deinterlace_enabled() {
            passes.push(PassType::Deinterlace);
        }
        if self.descratch.enabled {
            passes.push(PassType::DeScratch);
        }
        if self.noise_reduction.enabled {
            passes.push(PassType::NoiseReduction);
        }
        if self.dehalo.enabled {
            passes.push(PassType::Dehalo);
        }
        if self.deblock.enabled {
            passes.push(PassType::Deblock);
        }
        if self.deband.enabled {
            passes.push(PassType::Deband);
        }
        if self.sharpen.enabled {
            passes.push(PassType::Sharpen);
        }
        if self.chroma_fixes.enabled {
            passes.push(PassType::ChromaFixes);
        }
        if self.color_correction.enabled {
            passes.push(PassType::ColorCorrection);
        }
        if self.crop_resize.enabled && self.crop_resize.resize_enabled {
            // Resize (post-processing) - if not already added for crop
            if !passes.contains(&PassType::CropResize) {
                passes.push(PassType::CropResize);
            }
        }

        passes
    }

    /// Check if deinterlacing is enabled.
    fn deinterlace_enabled(&self) -> bool {
        self.deinterlace.enabled
    }

    /// Get count of enabled passes.
    pub fn enabled_pass_count(&self) -> usize {
        let mut count = 0;
        if self.deinterlace.enabled { count += 1; }
        if self.descratch.enabled { count += 1; }
        if self.noise_reduction.enabled { count += 1; }
        if self.dehalo.enabled { count += 1; }
        if self.deblock.enabled { count += 1; }
        if self.deband.enabled { count += 1; }
        if self.sharpen.enabled { count += 1; }
        if self.color_correction.enabled { count += 1; }
        if self.chroma_fixes.enabled { count += 1; }
        if self.crop_resize.enabled { count += 1; }
        count
    }

    /// Check if a specific pass is enabled.
    pub fn is_pass_enabled(&self, pass: PassType) -> bool {
        match pass {
            PassType::Deinterlace => self.deinterlace_enabled(),
            PassType::DeScratch => self.descratch.enabled,
            PassType::NoiseReduction => self.noise_reduction.enabled,
            PassType::Dehalo => self.dehalo.enabled,
            PassType::Deblock => self.deblock.enabled,
            PassType::Deband => self.deband.enabled,
            PassType::Sharpen => self.sharpen.enabled,
            PassType::ColorCorrection => self.color_correction.enabled,
            PassType::ChromaFixes => self.chroma_fixes.enabled,
            PassType::CropResize => self.crop_resize.enabled,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_default_pipeline() {
        let pipeline = ProcessingPipeline::default();
        assert!(!pipeline.noise_reduction.enabled);
        assert!(!pipeline.color_correction.enabled);
        assert!(!pipeline.chroma_fixes.enabled);
        assert!(!pipeline.crop_resize.enabled);
    }

    #[test]
    fn test_enabled_passes() {
        let mut pipeline = ProcessingPipeline::default();
        pipeline.noise_reduction.enabled = true;
        pipeline.color_correction.enabled = true;

        let passes = pipeline.enabled_passes();
        assert!(passes.contains(&PassType::Deinterlace));
        assert!(passes.contains(&PassType::NoiseReduction));
        assert!(passes.contains(&PassType::ColorCorrection));
        assert!(!passes.contains(&PassType::ChromaFixes));
    }

    #[test]
    fn test_serialization() {
        let pipeline = ProcessingPipeline::default();
        let json = serde_json::to_string(&pipeline).unwrap();
        assert!(json.contains("\"noiseReduction\""));
        assert!(json.contains("\"colorCorrection\""));
    }
}
